pallets/migrations/src/lib.rs

```diff
@@ -162,22 +157,27 @@ pub mod pallet {
 		let mut weight: Weight = 0u64.into();
 		let mut done: bool = true;
 
-		for migration in &MIGRATIONS {
+		for migration in &T::MigrationsList::get() {
 
 			// let migration_name = migration.friendly_name();
 			let migration_name = migration.friendly_name();
 			log::trace!("evaluating migration {}", migration_name);
 
-			let migration_state = <MigrationState<T>>::get(migration_name)
+			let migration_state = <MigrationState<T>>::get(migration_name.as_bytes())
 				.unwrap_or(Perbill::zero());
 
 			if migration_state < Perbill::one() {
 
+				// TODO: we don't currently have a reliable way to know "started"
+				// TODO: multiple calls to as_bytes() or to_vec() may be expensive
+				<Pallet<T>>::deposit_event(Event::MigrationStarted(migration_name.as_bytes().to_vec()));
```

Comment:
```rust
let migration_state = match MigrationState::<T>::get(migration_name.as_bytes() {
  None => {
    <Pallet<T>>::deposit_event(Event::MigrationStarted(migration_name.as_bytes().to_vec()));
    Perbill::zero()
  }
  Some(s) => s
}
```

Reply:
This TODO and your comment are related. There's a bit of a design decision here: do we leave it up to each pallet to track its own progress, or do we provide some helpers?

More fundamentally, can we provide generic helpers (e.g. progress is x of y) or would doing so make too many assumptions to be useful anyway? "x" and "y" may make no sense to a pallet, or at least may not map well to the work it needs to do. For example, a pallet might migrate some per-user storage, but then might have a heavy clean-up to do at the very end. Given that it must work within weight constraints, that might force it to use some sort of state machine (sounds like a nightmare...)

Additionally, there should be some goal of minimizing storage. A simple `bool` or `Perbill` is minimal, but virtually useless for helping a pallet.

I'm tempted to blow this out into a struct, something excessive would look like:

```rust
struct MigrationProgress {
    started: BlockNumber, // I believe we don't actually have this in on_runtime_upgrade, but let's pretend
    last_updated: BlockNumber,
    completed: Option<BlockNumber>,
    total_items: u64,
    completed_items: u64,
    total_weight: Weight, // why not
}
```

Most of these things can be inferred from the blockchain history (esp. with useful events) so I'm erring on simple for now. However, the main goal of this pallet is to make migrations easier, so it would be great if pallets didn't need to include additional storage/logic just to track their migration status.

Reply:
Yeah, I see what you mean. One option would be to introduce a trait for types that track progress. Then each migration could use its own struct that tracks progress in a way that makes sense for its pallet. But there is still a common interface that this migration manager pallet can rely on.

```rust
pub trait MigrationProgress {
  fn summarize_progress() -> Perbill;
```

Reply:
Would this trait interact with storage in some way?